internal/cmd/gocdk/serve.go

```diff
@@ -45,39 +51,70 @@ func serve(ctx context.Context, pctx *processContext, args []string) error {
 	}
 
 	// Check first that we're in a Go module.
-	moduleRoot, err := findModuleRoot(ctx, pctx.workdir)
+	var err error
+	opts.moduleRoot, err = findModuleRoot(ctx, pctx.workdir)
 	if err != nil {
 		return xerrors.Errorf("gocdk serve: %w", err)
 	}
 
 	// Verify that biome configuration permits serving.
-	biomeConfig, err := readBiomeConfig(moduleRoot, *biome)
+	biomeConfig, err := readBiomeConfig(opts.moduleRoot, opts.biome)
 	if xerrors.As(err, new(*biomeNotFoundError)) {
 		// TODO(light): Keep err in formatting chain for debugging.
 		return xerrors.Errorf("gocdk serve: biome configuration not found for %s. "+
 			"Make sure that %s exists and has `\"serve_enabled\": true`.",
-			*biome, filepath.Join(findBiomeDir(moduleRoot, *biome), biomeConfigFileName))
+			opts.biome, filepath.Join(findBiomeDir(opts.moduleRoot, opts.biome), biomeConfigFileName))
 	}
 	if err != nil {
 		return xerrors.Errorf("gocdk serve: %w", err)
 	}
 	if biomeConfig.ServeEnabled == nil || !*biomeConfig.ServeEnabled {
 		return xerrors.Errorf("gocdk serve: biome %s has not enabled serving. "+
 			"Add `\"serve_enabled\": true` to %s and try again.",
-			*biome, filepath.Join(findBiomeDir(moduleRoot, *biome), biomeConfigFileName))
+			opts.biome, filepath.Join(findBiomeDir(opts.moduleRoot, opts.biome), biomeConfigFileName))
 	}
 
-	// Start main serve loop.
+	// Start reverse proxy on address.
+	proxyListener, err := net.Listen("tcp", opts.address)
+	if err != nil {
+		return xerrors.Errorf("gocdk serve: %w", err)
+	}
+	myProxy := new(serveProxy)
+	group, groupCtx := errgroup.WithContext(ctx)
+	group.Go(func() error {
+		return runHTTPServer(groupCtx, proxyListener, myProxy)
+	})
+	proxyListener.Addr().String()
+
+	// Start main build loop.
 	logger := log.New(pctx.stderr, "gocdk: ", log.Ldate|log.Ltime)
-	if err := serveLoop(ctx, pctx, logger, *address, moduleRoot, *biome); err != nil {
+	group.Go(func() error {
+		return serveBuildLoop(groupCtx, pctx, logger, myProxy, proxyListener.Addr().(*net.TCPAddr), opts)
+	})
+	if err := group.Wait(); err != nil {
 		return xerrors.Errorf("gocdk serve: %w", err)
 	}
 	return nil
 }
 
-func serveLoop(ctx context.Context, pctx *processContext, logger *log.Logger, address string, moduleRoot, biome string) error {
+type serveOptions struct {
+	moduleRoot string
+	biome      string
+	address    string
+	basePort   int
+}
+
+// serveBuildLoop builds and runs the user's server and sets the proxy's
+// backend whenever a new built server becomes healthy. This loop will continue
+// until ctx's Done channel is closed. serveBuildLoop returns an error only
+// if it was unable to start the main build loop.
+func serveBuildLoop(ctx context.Context, pctx *processContext, logger *log.Logger, myProxy *serveProxy, proxyAddr *net.TCPAddr, opts *serveOptions) error {
```

Comment:
It looks like `proxyAddr` is only used as a string for output, maybe just pass it as a string?

Reply:
(avoiding the kind-of-weird cast to `*net.TCPAddr`)

Reply:
Done. (Albeit still with the type conversion.)